File: 05-Algorithms/code/python/seb_alg/topological_sort.py
"""Mirrors `typescript/src/topologicalSort.ts` exactly."""

from __future__ import annotations

import bisect
from typing import List, Sequence

from .errors import AlgoError
from .graph_common import build_adj_list


def topological_sort(n: int, edges: Sequence[Sequence[int]]) -> List[int]:
    """Kahn's algorithm (BFS by in-degree). Among all zero-in-degree
    vertices at any step, the lowest id is always removed first, making the
    order fully deterministic. Raises AlgoError("cycle") if not a DAG.
    O(V + E log V) due to the sorted-array tie-breaking structure.
    """
    adj = build_adj_list(n, edges, True)
    indegree = [0] * n
    for lst in adj:
        for v in lst:
            indegree[v] += 1

    available: List[int] = sorted(v for v in range(n) if indegree[v] == 0)

    order: List[int] = []
    while available:
        u = available.pop(0)
        order.append(u)
        for v in adj[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                bisect.insort(available, v)

    if len(order) != n:
        raise AlgoError("cycle", "graph contains a cycle; no topological order exists")
    return order

File: 05-Algorithms/code/typescript/src/topologicalSort.ts
import { AlgoError } from "./errors.js";
import { buildAdjList, type Edge } from "./graphCommon.js";

/**
 * Kahn's algorithm (BFS by in-degree) over a directed graph on n vertices.
 * Among all vertices with in-degree 0 at any given step, the lowest id is
 * always removed first, which makes the returned order fully deterministic
 * (not just "any valid topological order"). Throws AlgoError("cycle") if
 * the graph is not a DAG (fewer than n vertices get emitted). O(V + E log V)
 * due to the tie-breaking min-priority queue (a simple sorted-array stand-in
 * is used here since n is small in this lab; a real binary heap would drop
 * this to O(V + E)).
 */
export function topologicalSort(n: number, edges: Edge[]): number[] {
  const adj = buildAdjList(n, edges, true);
  const indegree = new Array(n).fill(0);
  for (const list of adj) for (const v of list) indegree[v]++;

  const available: number[] = [];
  for (let v = 0; v < n; v++) if (indegree[v] === 0) available.push(v);
  available.sort((a, b) => a - b);

  const order: number[] = [];
  while (available.length > 0) {
    const u = available.shift() as number;
    order.push(u);
    for (const v of adj[u]) {
      indegree[v]--;
      if (indegree[v] === 0) {
        const pos = available.findIndex((x) => x > v);
        if (pos === -1) available.push(v);
        else available.splice(pos, 0, v);
      }
    }
  }

  if (order.length !== n) throw new AlgoError("cycle", "graph contains a cycle; no topological order exists");
  return order;
}
